dpks/fdr.py

```diff
@@ -0,0 +1,231 @@
+                loc=feature_mean,
+                scale=feature_std,
+                size=self.n_samples,
+            )
+
+            self.decoy_features[:, i] = decoy_feature
+
+
+@numba.jit(nopython=True)
+def _fast_distribution_q_value(target_values, decoy_values, pit):  # type: ignore
+    target_area = np.trapz(target_values)
+
+    decoy_area = np.trapz(decoy_values)
+
+    if decoy_area == 0.0 and target_area == 0.0:
+        q_value = 0.0
+
+    else:
+        decoy_area = decoy_area * pit
+
+        q_value = decoy_area / (decoy_area + target_area)
+
+    return q_value
+
+
+def _fast_distribution_q_values(scores, target_function, decoy_function, pit):  # type: ignore
+    q_values = np.ones((len(scores),), dtype=np.float64)
+
+    max_score = np.max(scores)
+
+    for i in range(len(scores)):
+        integral_bounds = np.arange(scores[i], max_score, 0.1)
+
+        target_data = np.interp(integral_bounds, target_function[0], target_function[1])
+
+        decoy_data = np.interp(integral_bounds, decoy_function[0], decoy_function[1])
+
+        q_value = _fast_distribution_q_value(target_data, decoy_data, pit)
+
+        q_values[i] = q_value
+
+    return q_values
+
+
+class ScoreDistribution:
+    pit: float
+    X: np.ndarray
+    y: np.ndarray
+    target_spline: Tuple[np.ndarray, np.ndarray]
+    decoy_spline: Tuple[np.ndarray, np.ndarray]
+    target_scores: np.ndarray
+    decoy_scores: np.ndarray
+
+    def __init__(self, pit: float = 1.0, num_threads: int = 1):
+        numba.set_num_threads(num_threads)
+
+        self.pit = pit
+
+    def fit(self, X: np.ndarray, y: np.ndarray) -> ScoreDistribution:
+        self.X = X
+        self.y = y
+
+        self._estimate_bin_number()
+
+        target_indices = np.argwhere(y == 1)
+        decoy_indices = np.argwhere(y == 0)
+
+        self.target_scores = X[target_indices]
+        self.decoy_scores = X[decoy_indices]
+
+        self.target_spline = self._fit_function(self.target_scores)
+        self.decoy_spline = self._fit_function(self.decoy_scores)
+
+        return self
+
+    def min(self) -> float:
+        return float(np.min(self.X))
+
+    def max(self) -> float:
+        return float(np.max(self.X))
+
+    def _estimate_bin_number(self) -> None:
+        hist, bins = np.histogram(self.X, bins="auto")
+
+        self.num_bins = (bins[1:] + bins[:-1]) / 2
+
+    def _fit_function(self, scores: np.ndarray) -> Tuple[np.ndarray, np.ndarray]:
+        hist, bins = np.histogram(scores, bins=self.num_bins)
+
+        bin_centers = (bins[1:] + bins[:-1]) / 2
+
+        return bin_centers, hist
+
+    def q_values(self, X: np.ndarray) -> np.ndarray:
+        return _fast_distribution_q_values(
+            X, self.target_spline, self.decoy_spline, self.pit
+        )
+
+
+class DecoyCounter:
+    pit: float
+
+    def __init__(self, num_threads: int = 1, pit: float = 1.0) -> None:
+        numba.set_num_threads(num_threads)
+
+        self.pit = pit
```

Comment:
what is pit?

Reply:
"Percent Of Incorrect Targets", it is a measure used to down-wight the number of decoys used to calculate q-values. It's estimated by the percentage of the decoy distribution that is also covered by a labeled target distribution

Reply:
Okay. I think it'd be useful to name this something more telling, like `percent_incorrect_targets`. I will definitely forget what the variable is otherwise hehe

Reply:
Changed